rfd/0112-DB-credential-providers.md

```diff
@@ -0,0 +1,71 @@
+---
+authors: Sai Sandeep Rangisetti (sandeep.sai@flipkart.com)
+state: draft
+---
+
+# RFD 112 - Use a credential provider for DBs
+
+## What
+As of now all the self-hosted databases require enabling certificate authentication in the backend. This RFD proposes
+a way to avoid that requirement. Based on the labels we can decide to fetch credentials from a web service instead of
+using certificates for authentication. When this label is applied we will consider username as a token and exchange it
+for credentials with an external credential provider
+
+## Why
+Enabling certificate authentication requires config update and restart of the mysql server. This will not be scalable in
+large organizations which will have large number of database clusters. Instead, orgs can have a credential provider
+service like Vault which can give the credentials on need basis based on username and database. 
+
+## Detail
+### Overview 
+```
++--------------+               +--------------------+               +------------------+
+|              | pass token as |                    | pass exhanged |                  |
+|     user     +-------------->|      teleport      +-------------->|     database     |
+|              |   username    |                    |  credentials  |                  |
++--------------+               +--------+-----------+               +------------------+
+                                        |  ^
+                                        |  |
+                                        |  | Exchange token with credentials
+                                        V  |
+                               +-----------+--------+
+                               |                    |
+                               |     credential     |
+                               |      provider      |
+                               |                    |
+                               +--------------------+
+
+```
+
+### Get the details of the credential provider
+To exchange the username/token with actual credentials we need to read the required details like url, authentication
+mechanism and other details. We can read the details from the config file like below 
+```yaml
+db_service:
+  token:
+    url_tmpl: http://localhost:9000/database/{{.DBName}}/user/{{.Username}}/token/{{.Token}}
+    connection_timeout: 1s
+    read_timeout: 1s
+    tls:
+      insecure: false
+      # ca path is required if server is cert is self signed
+      ca_path: /tmp/ca.pem
+      # cert and key path required if authentication is based on mutual tls
+      cert_path: /tmp/cert.pem
+      key_path: /tmp/key.pem
+    authentication:
```

Comment:
Could Teleport use short-lived credentials to authenticate with the upstream credential provider? e.g in the case of Vault we could issue an OIDC ID Token as long as Vault was configured to trust Teleport's ID token issuer. @ryanclark has done some work recently that explored using OIDC ID tokens as a short-lived way for Teleport to authenticate with external services (in his case, AWS).

The other option for short-lived would be to allow this to use a certificate issued by the Teleport CA, and then to configure the credential provider to trust the Teleport CA.

Reply:
Teleport's ID token issuer is probably something like Okta or Google right? I think we can add one more authentication scheme called IDTOKEN which sends ID token issued to teleport as authentication. I will update the RFD with the changes